_posts/2024-09-20-reusing-rustc-components.md

```diff
@@ -0,0 +1,123 @@
+
+```rust
+println!("Hello {ferris}!")
+```
+
+the parser will emit three "pieces" - two for the static strings `"Hello "` and `"!"`, and one for the named argument `ferris`. Slightly simplified, the resulting vector will look like this:
+
+```rust
+vec![
+    Piece::Static("Hello "),
+    Piece::ArgumentNamed(ferris, Trait::Display),
+    Piece::Static("!"),
+]
+```
+
+Once we get these pieces back into the compiler, we can generate Rust code to call into `core::fmt` functions, so that at runtime your code will correctly call into the proper formatting traits like `Display`, `Debug` and so on. For the version of the language we are targeting, that generated code looks something like this:
+
+```rust
+io::_print(
+    fmt::Arguments::new_v1(
+        &["Hello, ", "!\n"],
+        &[fmt::ArgumentV1::new(&ferris, fmt::Display::fmt)]
+    )
+);
+```
+
+If you're thinking "Arthur, why didn't you just rewrite the parser, that sounds easy" well the answer is I am absolutely ass at writing parsers in C++. I need [`nom`](https://crates.io/crates/nom) or I can't function.
+
+Regardless, `rustc_parse_format` is a tiny `rustc` component, which in truth is not terribly difficult to implement. But tweaking the build system to allow the reuse of that component, as well as writing the various Rust <-> C++ interfaces needed for interacting with the `crate`, helped pave the way for the most crucial of the components we are currently reusing - `polonius`, the next generation Rust borrow-checker, which you can read more about [here](https://rust-lang.github.io/polonius/). It can already be used on `rustc` with the nightly `-Z polonius` option. It is currently being rewritten, and not stable yet, but will eventually allow a couple more constructs which are rejected by the current borrow-checking algorithm while still being just as correct.
+
+We plan on adding more components to `rustc` in the future where _NOTE: Reword "it makes sense"_. For example, there are currently efforts towards making a new trait solver for `rustc` - if separate enough from the compiler, it could be integrated and used by `gccrs` to perform trait-solving on our internal representation. Similarly, [Jack Wrenn](https://github.com/jswrenn) gave a fantastic talk at RustConf 2024 detailling checked-transmutes _NOTE: Add links to proposals, blogpost, etc_. Talking with Jack after his talk revealed that the algorithm powering the project is quite separate from the compiler. It uses its own graph representation, which is built from `rustc`'s internal representation, but could also be built using `gccrs`'!
```

Comment:
Same here. If you were going to do it, then these comments at least will act as reminders :)